ops/model.py

```diff
@@ -3704,20 +3640,24 @@ def _is_relation_not_found(model_error: Exception) -> bool:
         return 'relation not found' in str(model_error)
 
     def relation_ids(self, relation_name: str) -> list[int]:
-        relation_ids = self._run('relation-ids', relation_name, return_output=True, use_json=True)
-        relation_ids = typing.cast('Iterable[str]', relation_ids)
+        try:
+            with self._tracing_span('relation-ids', relation_name):
+                relation_ids = hookcmds.relation_ids(relation_name)
+        except hookcmds.Error as e:
+            self._check_for_security_event('relation-ids', e.returncode, e.stderr)
+            raise ModelError(e.stderr) from e
         return [int(relation_id.split(':')[-1]) for relation_id in relation_ids]
 
     def relation_list(self, relation_id: int) -> list[str]:
         try:
-            rel_list = self._run(
-                'relation-list', '-r', str(relation_id), return_output=True, use_json=True
-            )
-            return typing.cast('list[str]', rel_list)
-        except ModelError as e:
-            if self._is_relation_not_found(e):
+            with self._tracing_span('relation-list', relation_id):
+                return hookcmds.relation_list(relation_id)
+        except hookcmds.Error as e:
+            self._check_for_security_event('relation-list', e.returncode, e.stderr)
+            err = ModelError(e.stderr)
+            if self._is_relation_not_found(err):
```

Comment:
This is a bit awkward. Could we change the `_is_relation_not_found` and similar helpers just to take the string and pass `e.stderr` directly, and then `raise ModelError` below?